### ISSUE TITLE
How to calculate vote totals

### ISSUE THREAD

**Issue by whoabuddy** [whoabuddy/sip015.xyz#1]
Voting can occur through three methods and will occur across two reward cycles (currently: 46/47):

- solo stackers sending BTC
- delegated stackers sending STX
- liquid STX holders through stx.eco

For the first two, casting both a yes and no vote will invalidate the vote. At least 80M STX must participate, and 80% of the total must vote "yes".

For the third, there is no participation threshold but 66% of the total must vote "yes".

> In all cases, if a user votes at all, they can only vote in one of the two reward cycles. If a user casts a vote in both reward cycles, even if it is for the same decision, then the vote is discarded.

Each method also has a few ways they can interact, and more [details and scenarios](https://github.com/stacksgov/sips/blob/feat/sip-015/sips/sip-015/sip-015-network-upgrade.md#activation) are covered in the SIP.

If we can nail down the values for at least the first two if not all three votes, then a vote card could be displayed with total votes and on hover show a breakdown of votes per method. This would likely require some light caching as well to avoid having to tabulate the data live on every load.

**Comment by wileyj** [comment]
yea, it's a challenging problem without runnning some processes in the background to query/store the values. 
one thing i've been looking at is a "cron" service in vercel for something unrelated - that could be used to run a process to calculate the totals, then store them in a local json or redis (with some logic to delete/pop results that vote both yes and no)

basically this background process would run as an api endpoint, when the separate cron service is triggered every x minutes, it sends a request to this api endpoint. the code runs and stores the data. 

frontend then just queries the datastore to display it.

**Comment by wileyj** [comment]
i am partial to the local json idea (low barrier to entry for others running the same app), but depending on the amount of data to store/calculate that might be problematic

**Comment by whoabuddy** [comment]
I can create a cron-type service with CF workers and the CF key-value store, then serve that back to this app (or expose it as an API endpoint). If we map out the logic required between the voting methods happy to take that on as I've been itching to try this out as a caching layer for other Stacks apps.

**Comment by whoabuddy** [comment]
It'd also be open source so anyone could clone this repo and deploy their own with a $5 CF workers account (or possibly on the free tier, depending on how aggressive the cron job is).

**Comment by whoabuddy** [comment]
Thinking about this some more - the first layer would likely be indexing all the transactions to both the yes/no addresses between BTC/STX. From there, we can apply logic to determine valid votes - then total those to display the totals across method 1 + 2. 

For method 3 - @radicleart will there be something we can tap into on the stx.eco vote contract to determine total vote counts?

**Comment by mijoco-btc** [comment]
stx.eco has a similar api (docker based) which indexes the contracts every 3 minutes (cron like scheduler) and parses the data into mongodb. This allows it to supply voting history and fetch running totals / display the count real time (within 3 minute tolerance).

**Comment by mijoco-btc** [comment]
Just dug out the end point..

Passing in the contractId of the proposal

https://stx.eco/daoapi/v2/events/proposal/SP3JP0N1ZXGASRJ0F7QAHWFPGTVK9T2XNXDB908Z.edp001-adjust-dao-parameters

Returns a list of vote events - the ui parses to get the vote count.

```"voteEvent": {
"amount": 1000000,
"event": null,
"proposal": "SP3JP0N1ZXGASRJ0F7QAHWFPGTVK9T2XNXDB908Z.edp001-adjust-dao-parameters",
"voter": "SP2E3P9B712YSFQF5BAAA2D5G6RC8RQ0WNB0RRKM5",
"for": true
}
```

**Comment by whoabuddy** [comment]
That's great! That will give us access to the info we need, and I think we can do a bigger callout for Method 3 as well, it's sort of buried in that first paragraph and would be good to direct people to the correct method for their use case. Will open a separate issue for that.

Also linking the examples from SIP-015, which cover the overlaps in each method, this would be good to summarize so we make sure our logic follows the same calculation.
https://github.com/stacksgov/sips/blob/feat/sip-015/sips/sip-015/sip-015-network-upgrade.md#examples-1

**Comment by whoabuddy** [comment]
Following the [forum discussion here](https://forum.stacks.org/t/suggestion-for-a-fairer-stacks-2-1-vote/13983/4), instead of displaying vote totals we could display the amount of Stacked votes that count toward the threshold of 80M STX for the vote to pass using methods 1 and 2.

This gives a visual indicator that votes are coming in, and it would encourage participation if the threshold isn't met as the deadline closes.

### Implementation Notes

The Chakra UI provides a [simple component for progress bars](https://chakra-ui.com/docs/components/progress/usage).

Workers give us the ability to run functions that do these type of calculations, KV storage can help cache it, and we would just need to create endpoints for the data.

I'm going to think through the structure of this some more, but before I get too deep into it, @wileyj would you say the methods below cover the criteria for valid votes in the third layer? Anything else you can think of to change/add?

### Method 1

First layer: collecting the transactions to review

- get all TX from the yes and no BTC addresses (track by TXID)
- discard any that occur before the start height
- discard any that occur after the end height
- store the txs for each address for the next step
  - note: look at how fast this will grow / exceed KV limits
  - maybe slim down to a defined structure with just relevant data?

Second layer: breaking it down per address

- for both outcomes, evaluate every TX and:
  - record the input address for the tx
  - record the number of votes for that address
  - query and record the stacked amount for that reward address
  - set variable if vote is valid or not (based on number of votes)

Third layer: calculating the vote totals against the threshold

- threshold will be 80M Stacked STX participating per the SIP
- for each address from the second layer:
  - if the vote is valid, add it to the total

### Method 2

First layer: collecting the transactions to review

- get all TX from the yes and no STX addresses (track by TXID)
- discard any that occur before the start height
- discard any that occur after the end height
- store the txs for each address for the next step
  - note: look at how fast this will grow / exceed KV limits
  - maybe slim down to a defined structure with just relevant data?

Second layer: breaking it down per address

- for both outcomes, evaluate every TX and:
  - record the sending STX address for the TX
  - record the number of votes for that address
  - query and record the delegated stacked amount for that Stacks address
  - set variable if vote is valid or not (based on number of votes)

Third layer: calculating the vote totals against the threshold

- threshold will be 80M Stacked STX participating per the SIP
- for each address from the second layer:
  - if the vote is valid, add it to the total

**Comment by wileyj** [comment]
> I'm going to think through the structure of this some more, but before I get too deep into it, @wileyj would you say the methods below cover the criteria for valid votes in the third layer? Anything else you can think of to change/add?

the methods look good to me, method 1 seems like the route we want to take as well unless there are specific plans on how to handle delegate voting (i.e. a pool voting for all their participants as a single vote or a weighted vote)

**Comment by whoabuddy** [comment]
Yes, the activation section was revised to simplify the voting procedures, [see here for the full text](https://github.com/stacksgov/sips/blob/feat/sip-015/sips/sip-015/sip-015-network-upgrade.md#activation) and this snippet from voting method 1.

> Note that this voting procedure does not apply to Stacking pool operators. Stacking pool operator votes will not be considered.

Method 1 listed above captures all the solo stacking data, method 2 captures all the delegated stacking data, and the votes from both would be combined for the total Stacker vote.

**Comment by whoabuddy** [comment]
Making some progress on this one! For tabulating Method 2 (pooled stacking vote), the repo [whoabuddy/stacks-indexer](https://github.com/whoabuddy/stacks-indexer) contains a worker that acts as a caching layer for the vote transactions.

Some example links below, all that's left is getting the stacked amount from the PoX contract and making sure the duplicate votes aren't counted.

- [Main landing page with route descriptions](https://indexer.sip015.xyz/)
- [All yes votes](https://indexer.sip015.xyz/SP00000000000003SCNSJTCHE66N2PXHX)
- [All no votes](https://indexer.sip015.xyz/SP00000000000000DSQJTCHE66XE1NHQ)
- [Data stored in KV](https://indexer.sip015.xyz/data)
- [Vote tabulation - WIP](https://indexer.sip015.xyz/votes)

The end goal is to return JSON from this endpoint which can be consumed by a hook added in #31.

**Comment by mijoco-btc** [comment]
@whoabuddy - am seeing if I can make an independent count for this.

Wondering if there's a difference between calling SP000000000000000000002Q6VF78.pox/get-stacker-info(principal)  (at-block-height) and calling the end point.. https://stacks-node-api.mainnet.stacks.co/extended/v1/address/SPA490Y6V4XSD8BGT2YVBYESS6P7ZMK1ZTHN3BTE/stx?until_block=83985 and what the block height should be ?

**Comment by whoabuddy** [comment]
I believe we should use the block height that cycle 46 started: 762,550 / 82,914.

Right now I have the voting window set based on Bitcoin and Stacks blocks:

- Start: 762,550 / 82,914
- End: 766,750 / 87,114

That ending Stacks block might shift a bit depending on flash blocks during the two cycles.

~~As far as the two endpoints, if I'm understanding it correctly, the STX balance returned from the API would be the *unlocked balance* whereas the PoX call would show the *locked balance*.~~ I was wrong!
